test/test-agent.cpp

```diff
@@ -0,0 +1,69 @@
+#include <gtest/gtest.h>
+
+#include "agent.h"
+
+TEST(AgentTest, Basic)
+{
+    Rep::Agent agent = Rep::Agent().allow("/").disallow("/foo");
+    EXPECT_EQ(agent.directives().size(), 2);
+}
+
+TEST(AgentTest, ChecksAllowed)
+{
+    Rep::Agent agent = Rep::Agent().allow("/path");
+    EXPECT_TRUE(agent.allowed("/path"));
+    EXPECT_TRUE(agent.allowed("/elsewhere"));
+}
+
+TEST(AgentTest, HonorsLongestFirstPriority)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/path").allow("/path/exception");
+    EXPECT_TRUE(agent.allowed("/path/exception"));
+    EXPECT_FALSE(agent.allowed("/path"));
+}
+
+TEST(AgentTest, RobotsTextAllowed)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/robots.txt");
+    EXPECT_TRUE(agent.allowed("/robots.txt"));
+}
+
+TEST(AgentTest, DisallowNone)
+{
+    Rep::Agent agent = Rep::Agent().disallow("");
+    EXPECT_TRUE(agent.allowed("/anything"));
+}
+
+TEST(AgentTest, EscapedRule)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/a%3cd.html");
+    EXPECT_FALSE(agent.allowed("/a<d.html"));
+    EXPECT_FALSE(agent.allowed("/a%3cd.html"));
+}
+
+TEST(AgentTest, UnescapedRule)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/a<d.html");
+    EXPECT_FALSE(agent.allowed("/a<d.html"));
+    EXPECT_FALSE(agent.allowed("/a%3cd.html"));
+}
+
+TEST(AgentTest, EscapedRuleWildcard)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/a%3c*");
+    EXPECT_FALSE(agent.allowed("/a<d.html"));
+    EXPECT_FALSE(agent.allowed("/a%3cd.html"));
+}
+
+TEST(AgentTest, UnescapedRuleWildcard)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/a<*");
+    EXPECT_FALSE(agent.allowed("/a<d.html"));
+    EXPECT_FALSE(agent.allowed("/a%3cd.html"));
+}
+
+TEST(AgentTest, AllowUrl)
```

Comment:
It's meant to test that it can accept a URL. Originally, I had meant to have two separate methods: `allowed` and `url_allowed`, the later of which accepts a full URL (rather than just a path).

However, in re-reading this, both have to go through the same pathway: `Url::Url(string).defrag().escape().fullpath()`, so I think I'll just collapse them into a single method.

I'll also rename this test to `AcceptsFullUrl`